### PR TITLE
Implement start/next/done for CircularDeque

### PR DESCRIPTION
**Pull request by kdmurray91**
Hi,

This PR simply adds the `start`/`next`/`done` methods required for idiomatic for-loop iteration over CircularDeques.

I'm a relative Julia novice, so any tips (especially regarding performance) are appreciated.

Cheers,
Kevin

### CODE REVIEW

File: test/test_circ_deque.jl [hunk 1]
```diff
@@ -78,6 +78,11 @@ end
     empty!(D)
     unshift!(D, 40)
     @test front(D) == back(D) == 40
+
+    # Test iteration over loop
+    D = CircularDeque{Int}(5)
+    for i in 1:5 push!(D, i) end
+    @test collect(i for i in D) == collect(1:5)
```
Comment (kmsquire) [inline test/test_circ_deque.jl]: Generators aren't available on Julia 0.4.  To maintain compatibility, you could enclose the generator inside an array:

```julia
collect([i for i in D])
```

File: src/circ_deque.jl [hunk 2]
```diff
@@ -80,6 +80,11 @@ end
     ret
 end
 
+# Iteration via getindex
+Base.start(d::CircularDeque) = 1
+Base.next(d::CircularDeque, i) = (d[i], i+1)
+Base.done(d::CircularDeque, i) = i == length(d) + 1
```
Comment (kmsquire) [inline src/circ_deque.jl]: Using `getindex` is a good first pass.  It does do a number of checks that might (?) not be necessary during iteration, so if you wanted, you could try to implement a more complicated version of these functions which access the buffer directly. Not sure if the actual benefit.

Comment (kdmurray91) [inline src/circ_deque.jl]: I've pushed an "enhanced" version that uses a new `unsafe_getindex` to avoid bounds checking via getindex.

**Comment by kmsquire** [comment]
Thanks for the contribution!

**Pull request merged**